### PR TITLE
Executor ITF: common base fixture

### PR DESCRIPTION
**Pull request by MBoldyrev**
Signed-off-by: Mikhail Boldyrev <miboldyrev@gmail.com>

### Description of the Change

`ExecutorTestBase` is a common base class for Executor ITF tests. It supports various backends as parameter.

`BasicExecutorTest` is an implementation of Executor ITF test class that gets backend parameters from Google test framework. It is supposed to be the used in most test cases. Though there are test cases where it is not suitable - the ones that require more complex parameters than just the Executor ITF backend. An example can be seen in my WIP branch:
https://github.com/hyperledger/iroha/blob/9ac7d8106eafdfac5c1adb826f4278c4b4630b27/test/integration/executor/query_permission_test.hpp#L37-L42

### Benefits

One more step towards the new shiny Executor ITF.

### CODE REVIEW

File: test/integration/executor/executor_fixture.hpp [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+#include "module/shared_model/mock_objects_factories/mock_query_factory.hpp"
+
+namespace executor_testing {
+
+  namespace error_codes {
+    using shared_model::interface::ErrorQueryResponse;
+
+    // TODO [IR-1816] Akvinikym 06.12.18: remove these constants after
+    // introducing a uniform way to use them in code
+    static constexpr ErrorQueryResponse::ErrorCodeType kNoStatefulError = 0;
```
Comment (luckychess) [inline test/integration/executor/executor_fixture.hpp]: Maybe introduce an Enum?

File: test/integration/executor/executor_fixture.hpp [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+    void addAsset(
+        const shared_model::interface::types::AccountIdType &dest_account_id,
+        const shared_model::interface::types::AssetIdType &asset_id,
+        const shared_model::interface::Amount &quantity);
+
+    //  ---------------- checkers -----------------
+
+    struct AssetQuantity {
+      template <typename... Types>
```
Comment (luckychess) [inline test/integration/executor/executor_fixture.hpp]: Looks a bit too complicated and could be simplified.

File: test/integration/executor/executor_fixture.cpp [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+ */
+
+#include "integration/executor/executor_fixture.hpp"
+
+#include <gtest/gtest.h>
+#include <boost/range/adaptor/transformed.hpp>
+#include "cryptography/crypto_provider/crypto_defaults.hpp"
+#include "framework/common_constants.hpp"
+#include "framework/result_gtest_checkers.hpp"
+#include "interfaces/query_responses/query_response.hpp"
```
Comment (luckychess) [inline test/integration/executor/executor_fixture.cpp]: It's already included in hpp file.

File: test/integration/executor/executor_fixture.cpp [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+  void checkCommandError(const CommandResult &command_result,
+                         CommandError::ErrorCodeType error_code) {
+    if (auto err = resultToOptionalError(command_result)) {
+      EXPECT_EQ(err->error_code, error_code);
+    } else {
+      ADD_FAILURE() << "Did not get the expected command error!";
+    }
+  }
+
+  std::pair<std::string, std::string> splitNameAndDomain(std::string id,
```
Comment (luckychess) [inline test/integration/executor/executor_fixture.cpp]: Suggest to use const reference (here and in functions below).

File: test/integration/executor/executor_fixture.cpp [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+      std::vector<decltype(make_asset_matcher(reference_quantities.front()))>>(
+      reference_quantities | boost::adaptors::transformed(make_asset_matcher));
+
+  EXPECT_THAT(test_quantities,
+              ::testing::UnorderedElementsAreArray(asset_matchers));
+}
+
+void ExecutorTestBase::checkAssetQuantities(
+    const std::string &account_id,
+    const std::vector<AssetQuantity> quantities) {
```
Comment (luckychess) [inline test/integration/executor/executor_fixture.cpp]: Probably forgotten &

File: test/integration/executor/executor_fixture.cpp [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+      *getItf().getMockCommandFactory()->constructAddAssetQuantity(asset_id,
+                                                                   quantity)));
+  assertResultValue(getItf().executeMaintenanceCommand(
+      *getItf().getMockCommandFactory()->constructTransferAsset(
+          kAdminId, dest_account_id, asset_id, "adding asset", quantity)));
+}
+
+void ExecutorTestBase::checkAssetQuantities(
+    const AccountAssetCollectionType &test_quantities,
+    const std::vector<AssetQuantity> reference_quantities) {
```
Comment (luckychess) [inline test/integration/executor/executor_fixture.cpp]: Reference plz

File: test/integration/executor/executor_fixture.hpp (hunk 1 above)
Comment (MBoldyrev) [reply test/integration/executor/executor_fixture.hpp]: decided that a desireable solution falls under the present todo

**Pull request merged**